### PR TITLE
Allow attach meta information using  block (#60, #79)

### PR DESCRIPTION
**Pull request by DmitryTsepelev**
This PR addresses #60 - now metadata is saved as a JSON field and there is no need to encode/decode it. I've implemented the API proposed by [here](https://github.com/palkan/logidze/issues/60#issuecomment-374819405):

```ruby
Logidze.with_meta(ip: request.ip) do
  post.save!
end
```

### CODE REVIEW

**Comment by DmitryTsepelev** [comment]
@palkan Thanks for the review!

I've added a couple of specs to check how nested `with_meta` blocks work, because each block is wrapped into its own transaction and a `LOCAL` keyword is used to store the meta inside the transaction. Does it suit our needs or I'm missing something?

**Comment by DmitryTsepelev** [comment]
Please ignore my previous question, my spec was wrong.

I've added handling for nested `with_meta` calls in a way that it merges the existing meta with the new one and then, after `yield`, reverts it back to the previous meta instead of using the default value. 

Also, I've changed `with_responsible` call to use the new functionality (and merged specs). `Version#responsible_id` supports both old and new data structures.

File: lib/logidze/meta.rb [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+          reset_meta_param
+          result
+        end
+      end
+
+      private
+
+      delegate :connection, to: ActiveRecord::Base
+
+      def prev_meta
```
Comment (palkan) [inline lib/logidze/meta.rb]: Looks a bit complicated.

Can't we just store the previous meta as a thread-local variable and avoid DB queries and parsing?

File: lib/generators/logidze/install/templates/migration.rb.erb [hunk 2]
```diff
@@ -9,6 +9,7 @@ class <%= @migration_class_name %> < ActiveRecord::Migration<%= ActiveRecord::VE
           BEGIN
           EXECUTE 'ALTER DATABASE ' || quote_ident(current_database()) || ' SET logidze.disabled=' || quote_literal('');
           EXECUTE 'ALTER DATABASE ' || quote_ident(current_database()) || ' SET logidze.responsible=' || quote_literal('');
```
Comment (palkan) [inline lib/generators/logidze/install/templates/migration.rb.erb]: Do we still need this `logidze.responsible` setting?

File: lib/logidze/meta.rb [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+module Logidze # :nodoc:
+  # Provide methods to attach meta information
+  module Meta
+    def with_meta(meta, &block)
+      MetaTransaction.wrap_with(meta, &block)
+    end
+
+    def with_responsible(responsible_id, &block)
+      return yield if responsible_id.nil?
+      meta = { Logidze::History::Version::RESPONSIBLE => responsible_id }
```
Comment (charlie-wasp) [inline lib/logidze/meta.rb]: Do we have a risk that users can inadvertently overwrite this key with their own meta? Yeah, I know that the usage of key `r` is unlikely, but hey, it's software development  :) Maybe `_r` or something would be safer?

File: lib/logidze/meta.rb (hunk 1 above)
Comment (charlie-wasp) [reply lib/logidze/meta.rb]: Using anything that involves threads doesn't sound simple to me at all 😅 Just my two cents

Maybe we can somehow use postgres `||` operator in `set_meta_param` method? It can merge two jsonb values. Although I had no luck applying it, maybe there is some postgres wizardry, that would allow us to do that.

File: lib/logidze/meta.rb (hunk 3 above)
Comment (DmitryTsepelev) [reply lib/logidze/meta.rb]: Yep, there are couple of options we could consider:

1. do nothing :)
2. use `_r` and hope that no one would override it
3. exclude `r` key (or whatever we end up with) from the meta hash silently
4. throw an exception when someone sets `r` meta key

File: lib/logidze/meta.rb (hunk 1 above)
Comment (palkan) [reply lib/logidze/meta.rb]: Threads are not evil)

Keeping the current value in the thread is the same as keeping it in the current postgres session in terms of potential problems) But thread is _local_

**Comment by DmitryTsepelev** [comment]
Thanks for the reviews!

Addressed all the comments, one commit per comment to make the review easier (I'll squash them later). Please let me know if solution with thread-local variables looks fine. Also, I've renamed `PgHelpers` back to the `Migration` cause I'm not using it anymore.

File: lib/generators/logidze/install/templates/migration.rb.erb (hunk 2 above)
Comment (DmitryTsepelev) [reply lib/generators/logidze/install/templates/migration.rb.erb]: nope, removed the line

File: lib/logidze/meta.rb [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+      private
+
+      def call_block_in_meta_context
+        meta_stack.push(meta)
+        pg_set_meta_param(current_meta)
+        result = block.call
+        pg_reset_meta_param
+        result
+      rescue StandardError => exception
```
Comment (palkan) [inline lib/logidze/meta.rb]: Let's use `begin ... ensure` instead

Comment (DmitryTsepelev) [reply lib/logidze/meta.rb]: I tried it initially to avoid memoizing and passing down the current meta at the beginning of the block, switching back

File: lib/logidze/meta.rb [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+        prev_meta = current_meta
+
+        meta_stack.push(meta)
+
+        pg_set_meta_param(current_meta)
+        result = block.call
+        pg_reset_meta_param(prev_meta)
+
+        result
+      ensure
```
Comment (palkan) [inline lib/logidze/meta.rb]: I don't see whether we have a spec for this case? We need to make sure that we pop from `meta_stack` when we're leaving `with_meta` block

Comment (DmitryTsepelev) [reply lib/logidze/meta.rb]: Added couple of checks to the existing specs and one separate spec to handle nested blocks. I wonder if there is a better way to test the same thing without letting the test know that we use `Thread.current`, maybe it worth trying to update `subject` in different places and verify that it's meta is correct

**Comment by DmitryTsepelev** [comment]
Squashed everything to a single commit, please let me know if we need to address anything else :)

File: CHANGELOG.md [hunk 6]
```diff
@@ -3,6 +3,7 @@
 ## master
 
 - [Fixes [#75](https://github.com/palkan/logidze/issues/70)] Fix association versioning with an optional belongs to ([@ankursethi-uscis][])
+- [PR [#79](https://github.com/palkan/logidze/pull/13)] Allow adding meta information to versions using `with_meta` (addressed [Issue [#60]](https://github.com/palkan/logidze/issues/60)). ([@DmitryTsepelev][])
```
Comment (palkan) [inline CHANGELOG.md]: Let's add an example here.

And also a note about how to upgrade form the older versions

Comment (DmitryTsepelev) [reply CHANGELOG.md]: Done, added a note about the change in the data structure, if someone works with the `version.data['r']` directly he'll have to make the fallback manually, but otherwise no actions needed to upgrade

Comment (palkan) [reply CHANGELOG.md]: > otherwise no actions needed to upgrade

Users have to regenerate logidze stored functions, don't they? Thus they have to create an upgrade migration https://github.com/palkan/logidze#upgrade-from-previous-versions

Comment (DmitryTsepelev) [reply CHANGELOG.md]: Oh, right, added it too

File: .rubocop.yml [hunk 7]
```diff
@@ -16,6 +16,7 @@ AllCops:
     - 'bench/**/*'
     - 'Rakefile'
     - 'Gemfile'
+    - 'CHANGELOG.md'
```
Comment (palkan) [inline .rubocop.yml]: Why did you add `CHANGELOG.md` to ignore?
